spark/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingReader.scala

```diff
@@ -583,6 +588,191 @@ trait DataSkippingReaderBase
+      // data skipping - simply remove the ineligible side.
+      case And(left, right) =>
+        val leftResult = rewriteDataFiltersAsPartitionLikeInternal(left, clusteringColumnPaths)
+        val rightResult = rewriteDataFiltersAsPartitionLikeInternal(right, clusteringColumnPaths)
+        (leftResult, rightResult) match {
+          case (Some((newLeft, statsLeft)), Some((newRight, statsRight))) =>
+            Some((And(newLeft, newRight), statsLeft ++ statsRight))
+          case _ => leftResult.orElse(rightResult)
+        }
+      // For any other expression, recursively rewrite the expression's children unless the
+      // expression might be very expensive to evaluate (JSON parsing, regex, etc.).
+      case other if !other.containsAnyPattern(INVOKE, JSON_TO_STRUCT, LIKE_FAMLIY,
+        REGEXP_EXTRACT_FAMILY, REGEXP_REPLACE) =>
+        val childResults = other.children.map(
+          rewriteDataFiltersAsPartitionLikeInternal(_, clusteringColumnPaths))
+        Option.whenNot (childResults.exists(_.isEmpty)) {
+          val (children, stats) = childResults.map(_.get).unzip
+          (other.withNewChildren(children), stats.flatten.toSet)
+        }
+      // For expensive expressions, we can't safely rewrite the expression.
+      case _ => None
+    }
+
+    /**
+     * Returns an expression that returns true if a file must be read because of a mismatched
+     * min-max value or partial nulls on a given column. For these files, it's not safe to apply
+     * arbitrary partition-like filters.
+     */
+    private def fileMustBeScanned(
+        resolvedPartitionLikeReference: ResolvedPartitionLikeReference,
+        numRecordsColOpt: Option[Column]): Expression = {
+      // Construct an expression to determine if all records in the file are null.
+      val nullCountExpr = resolvedPartitionLikeReference.nullCountExpr
+      val allNulls = numRecordsColOpt match {
+        case Some(physicalNumRecords) => EqualTo(nullCountExpr, physicalNumRecords.expr)
+        case _ => Literal(false)
+      }
+
+      // Note that there are 2 other differences in behavior between unpartitioned and partitioned
+      // tables:
+      // 1. If the column is a timestamp, the min-max stats are truncated to millisecond precision.
+      //    We shouldn't apply partition-like filters in this case, but
+      //    rewriteDataFiltersAsPartitionLikeInternal validates the column is not a Timestamp,
+      //    so we don't have to check here.
+      // 2. The min-max stats on a string column might be truncated for an unpartitioned table.
+      //    Note that just validating that the min and max are equal is enough to prevent this case
+      //    - if the string is truncated, the collected max value is guaranteed to be longer than
+      //    the min value due to the tiebreaker character(s) appended at the end of the max.
+      Not(
+        Or(
+          allNulls,
+          And(
+            EqualNullSafe(
```

Comment:
Shouldn't this be `EqualTo` in case the min/max stats are missing and inferred to be NULL?

Reply:
I think this _theoretically_ can be EqualNullSafe (if the min/max stats are missing and inferred to be NULL, then the `verifyStatsForFilter` check should trigger and we should read the file anyways). 

However, for clarity, we never expect the min-max stats to be null when in this expression branch (given the check below that there are no nulls on the column), so replaced with EqualTo.